### PR TITLE
Use export= for node:cluster module

### PR DESCRIPTION
**Pull request by ggoodman**
## Description

This PR attempts to fix the experience of using `require('cluster')` and `require('node:cluster')` in the context of a CommonJS module.

Prior to this PR, a CommonJS module importing this would have been misled into trying to use the `default` export to find the expected APIs.

The main concept was taken from a [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53669/files?show-viewed-files=true&file-filters%5B%5D=#r652256583) made by @rbuckton on the initial transition of `@types/node`'s `cluster` module to more ESM syntax.

### CODE REVIEW

**Comment by SimonSchick** [comment]
Can you provide tests and/or code examples that are problematic?

**Comment by ggoodman** [comment]
> Can you provide tests and/or code examples that are problematic?

@SimonSchick I'm struggling to find a pattern that would allow me to express a test that would show that the following gets marked as invalid by TypeScript with the current type definitions:

```ts
//@ts-check
const Cluster = require('cluster');

// The following will be treated as a type error
if (Cluster.isPrimary) {
  Cluster.fork();
}
```

Conversely, the status quo would allow the following to successfully type-check but would fail at runtime. Note that we're grabbing the `.default` property:

```ts
const Cluster = require('cluster').default;

// The following will successfully type check but will throw uncaught exceptions at runtime.
if (Cluster.isPrimary) {
  Cluster.fork();
}
```

**Comment by ggoodman** [comment]
I've adjusted the existing test slightly but now note a new error that is being produced that I suspect is some weird consequence of the hack used to define the `cluster.Worker` class. The following test now fails where it didn't before: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a3ae536413c86e8fbb392fd360a29cec8b1f201b/types/node/test/cluster.ts#L24

**Comment by ggoodman** [comment]
With a bit of gymnastics, I think we _appear_ to have a working solution.

**Comment by LinusU** [comment]
It seems like this changes all tests in this way:

```diff
- import cluster from 'node:cluster'
+ import * as cluster from 'node:cluster'
```

Does this mean that user code that uses `import cluster from 'node:cluster'` will break?

That is valid Node.js, and from what I've seen it's more common than doing `import * as`. The Node.js standard library exports things both individually, and as an object as the default export.

That is also the way the documentation teaches us to import:
https://nodejs.org/api/cluster.html#cluster

**Comment by ggoodman** [comment]
Hi @LinusU, I've just done a bunch of tests to see what works and does not work (without considering whether it type-checks).

**Node.js `"type": "module`**

```js
import * as cluster from 'cluster';

if (cluster.isPrimary) {
  cluster.fork();
} else {
  console.log('Fork succeeded');
}
```

✅  works

```js
import cluster from 'cluster';

if (cluster.isPrimary) {
  cluster.fork();
} else {
  console.log('Fork succeeded');
}
```

✅  works

**Transpiled to JS using `tsc` without `esModuleInterop`**

```json
{
  "compilerOptions": {
    "module": "CommonJS"
  },
  "include": ["src"]
}```

```ts
// src/index.ts
import cluster from 'cluster';

if (cluster.isPrimary) {
  cluster.fork();
} else {
  console.log('Fork succeeded');
}
```

Produces:

```js
"use strict";
exports.__esModule = true;
var cluster_1 = require("cluster");
if (cluster_1["default"].isPrimary) {
    cluster_1["default"].fork();
}
else {
    console.log('Fork succeeded');
}
```

❌ Does not work

```ts
// src/index.ts
import * as cluster from 'cluster';

//@ts-ignore
if (cluster.isPrimary) {
  //@ts-ignore
  cluster.fork();
} else {
  console.log('Fork succeeded');
}
```

Produces:

```js
"use strict";
exports.__esModule = true;
var cluster = require("cluster");
//@ts-ignore
if (cluster.isPrimary) {
    //@ts-ignore
    cluster.fork();
}
else {
    console.log('Fork succeeded');
}
```

✅  Works

However, enabling `"esModuleInterop"` allows this to work once again. 

My conclusion is that the status quo either:
1. Requires users to enable `esModuleInterop`; or
2. Write code that type checks and transpiles but will fail at runtime; or
3. Write code that doesn't type check but works.

I think that this PR addresses those concerns but am by no means an expert in the nuance of definition file authoring.

**Comment by LinusU** [comment]
My use case (which I think will be very common going forward seeing as the community seems to be shifting towards ESM) is `"type": "module"` in package.json, and `"module": "ES2020", "moduleResolution": "node"` in tsconfig. In this setup the following code compiles to the exact same code:

```typescript
import cluster from 'cluster';

if (cluster.isPrimary) {
  cluster.fork();
} else {
  console.log('Fork succeeded');
}
```

produces:

```js
import cluster from 'cluster';
if (cluster.isPrimary) {
    cluster.fork();
}
else {
    console.log('Fork succeeded');
}
```

I think it would be unfortunate if this stopped working...

> This PR attempts to fix the experience of using `require('cluster')` and `require('node:cluster')` in the context of a CommonJS module.

How are you writing your code, when you say require do you mean like this?

```typescript
import cluster = require('cluster')
```

Because that does indeed unfortunately type cluster as something like `{ default: ... }`, when that is not the case 😞 

It would be great if we could solve it so that both cases works!

**Comment by ggoodman** [comment]
I agree with your sentiment @LinusU and can see that it does indeed work as noted. I can see that it would break with the suggestions in this PR. It would probably be possible to add a `.default` property to the `Context` interface and to see if that tricks TypeScript into allowing a default import. However, doing so would result in allowing people to write TypeScript code that passes type checks but fails at runtime (when targeting CommonJS and not using `"esModuleInterop"`).

I think the suggested changes are a more accurate reflection of reality and likely provide a better experience for people who are authoring in CommonJS and/or targeting CommonJS. It is unfortunate that I can't think of a way to avoid sacrificing some ergonomics for early adopters in this case.

**Comment by ggoodman** [comment]
I found this in the docs but haven't found the 'how': https://nodejs.org/api/esm.html#builtin-modules

> [Core modules](https://nodejs.org/api/modules.html#core-modules) provide named exports of their public API. A default export is also provided which is the value of the CommonJS exports. The default export can be used for, among other things, modifying the named exports. Named exports of builtin modules are updated only by calling [module.syncBuiltinESMExports()](https://nodejs.org/api/module.html#modulesyncbuiltinesmexports).

**Comment by ggoodman** [comment]
The key reasons why this change should be seriously considered can be summarized by these current problems:

- ❌  ESM authors that use the default import and target `CommonJS` without `esModuleInterop` enabled will successfully type check but break at runtime.
- ❌  CJS authors can either write code that type checks but fails at runtime or write code that fails type checks but runs. `esModuleInterop` doesn't offer a solution here. **There is no way to write CJS code that type checks AND runs.**

The tradeoff we would need to accept is only one of ergonomics, not one of runtime correctness. With the changes in this PR, ESM authors and CJS authors alike would be able to write code that type checks and runs regardless of `esModuleInterop`, authoring module format or target module format. Some people will discover that code written as `import cluster from 'cluster'` will stop type-checking (but will continue running) and will need to change to `import * as cluster from 'cluster'`. Both syntaxes will result in code that can successfully run.

Given these arguments, I think balance of trade-offs weighs in favour of adopting this change.

File: types/node/v16/test/cluster.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import { connect } from 'net';
-import cluster, { ClusterSettings, Worker } from 'node:cluster'; // requires synthetic default imports, is this worth it?
+import * as cluster from 'node:cluster'; // requires synthetic default imports, is this worth it?
```
Comment (peterblazejewicz) [inline types/node/v16/test/cluster.ts:2]: why are you changing the imports, they should still work, e.g. this works with your pr branch:
```ts
import cluster  from 'node:cluster';
import type { ClusterSettings, Worker } from 'node:cluster';
```

Comment (ggoodman) [reply types/node/v16/test/cluster.ts:2]: I got some linting errors telling me I wasn't allowed to have two imports from the same source. Let me see if I can track it down.

I had initially used exactly the same syntax you propose but had the linter yelling at me.

Comment (ggoodman) [reply types/node/v16/test/cluster.ts:2]: Here is what I was seeing. IMHO, this lint rule is forcing us to use less-than-ideal syntax and is probably not adding the value we want.
```
2> Error: /Users/ggoodman/Projects/ggoodman/DefinitelyTyped/types/node/test/cluster.ts:3:1
2> ERROR: 3:1  no-duplicate-imports  Multiple imports from 'node:cluster' can be combined into one.
2> 
2>     at testTypesVersion (/Users/ggoodman/Projects/ggoodman/DefinitelyTyped/node_modules/@definitelytyped/dtslint/dist/index.js:194:15)
2>     at async runTests (/Users/ggoodman/Projects/ggoodman/DefinitelyTyped/node_modules/@definitelytyped/dtslint/dist/index.js:170:13)
```

Comment (peterblazejewicz) [reply types/node/v16/test/cluster.ts:2]: oh, yes, that's good code hygiene rule. On DT you can escape this using inline comment.
I understand, that's ok for the change.

**Comment by ggoodman** [comment]
@SimonSchick I wonder if you might be able to take a fresh look at this now that the tests have been adjusted and the build succeeds.

**Comment by ggoodman** [comment]
It appears that others are actually facing this issue as well: https://stackoverflow.com/q/69094343

**Comment by RyanCavanaugh** [comment]
@weswigham you seem most qualified to weigh in here

**Review by amcasey — CHANGES_REQUESTED** [review]
@ggoodman Thanks for approaching this change in such a thoughtful way.  I'm going to flag it as a Request Changes for now, not to express any opinion on its desirability but because I think a change this tricky requires affirmative sign-off from one or more owners.

**Comment by ggoodman** [comment]
I'd encourage the participants of this issue to revisit this. Letting the bot close this out is basically saying that we're OK keeping types for a node core module that will **mislead users into writing code that will crash at runtime**.

**Comment by mcollina** [comment]
I opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60424 to add myself to the owners.

**Pull request closed without merging**